ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -650,7 +650,8 @@ protected void mergeSchema(TableScanOperator discardableTsOp, TableScanOperator
 
   private static boolean compatibleSchema(TableScanOperator tsOp1, TableScanOperator tsOp2) {
     return tsOp1.getNeededColumns().equals(tsOp2.getNeededColumns())
-        && tsOp1.getNeededColumnIDs().equals(tsOp2.getNeededColumnIDs());
+        && tsOp1.getNeededColumnIDs().equals(tsOp2.getNeededColumnIDs())
+        && tsOp1.getConf().getVirtualCols().equals(tsOp2.getConf().getVirtualCols());
```

Comment:
is it possible to get NULLs here?

Reply:
Column pruning always set an empty array but I changed these to use `Obejcts.equals` for safety.